server/scsynth/SC_PortAudio.cpp

```diff
@@ -262,39 +268,198 @@ int SC_PortAudioDriver::PortAudioCallback(const void* input, void* output, unsig
+            && Pa_GetDeviceInfo(*inDevice)->hostApi != Pa_GetDeviceInfo(*outDevice)->hostApi) {
+            fprintf(stdout, "Requested devices %s and %s use different API. ", GetPaDeviceName(*inDevice).c_str(),
+                    GetPaDeviceName(*outDevice).c_str());
+            *outDevice = Pa_GetHostApiInfo(Pa_GetDeviceInfo(*inDevice)->hostApi)->defaultOutputDevice;
+            fprintf(stdout, "Setting output device to %s.\n", GetPaDeviceName(*outDevice).c_str());
+        }
+        // check for matching sampleRate or requested sample rate
+        if (*inDevice != paNoDevice && *outDevice != paNoDevice) {
+            PaStreamParameters in_parameters, out_parameters;
+            in_parameters = GetPaStreamParameters(inDevice, Pa_GetDeviceInfo(*inDevice)->maxInputChannels, 0);
+            out_parameters = GetPaStreamParameters(outDevice, Pa_GetDeviceInfo(*outDevice)->maxOutputChannels, 0);
+            if (sampleRate) {
+                // check if devices can support requested SR
+                PaError err = Pa_IsFormatSupported(&in_parameters, &out_parameters, sampleRate);
+                if (err != paNoError) {
+                    fprintf(stdout, "\nRequested sample rate %f for devices %s and %s is not supported.\n", sampleRate,
+                            GetPaDeviceName(*inDevice).c_str(), GetPaDeviceName(*outDevice).c_str());
+                    return err;
+                }
+            } else {
+                // if we don't request SR, check if devices have maching SR
+                uint32 inSR = uint32(Pa_GetDeviceInfo(*inDevice)->defaultSampleRate);
+                uint32 outSR = uint32(Pa_GetDeviceInfo(*outDevice)->defaultSampleRate);
+                if (inSR != outSR) {
+                    // if defaults are different, check if both devices can be opened using the OUTPUT's SR
+                    PaError err = Pa_IsFormatSupported(&in_parameters, &out_parameters,
+                                                       Pa_GetDeviceInfo(*outDevice)->defaultSampleRate);
+                    if (err != paNoError) {
+                        fprintf(stdout,
+                                "\nRequested devices %s and %s use different sample rates. "
+                                "Please set matching sample rates "
+                                "in the Windows Sound Control Panel and try again.\n",
+                                GetPaDeviceName(*inDevice).c_str(), GetPaDeviceName(*outDevice).c_str());
+                        return err;
+                    }
+                }
+            }
+        }
+
+        // in case we still don't have a proper device, use the default device
+        if (*inDevice == paNoDevice || *outDevice == paNoDevice) {
+            *inDevice = Pa_GetDefaultInputDevice();
+            *outDevice = Pa_GetDefaultOutputDevice();
+            if (*inDevice != paNoDevice && *outDevice != paNoDevice)
+                fprintf(stdout, "Selecting default system input/output devices\n");
```

Comment:
should there be an else here?

Reply:
I thought not.

If the device selection fails for some reason (and devices are still `paNoDevice`) then function continues/exits without setting devices. If that happens, then the [DriverSetup](https://github.com/dyfer/supercollider/blob/topic/scsynth-portaudio-input-output-devices/server/scsynth/SC_PortAudio.cpp#L459) continues to [OpenDefaultStream](https://github.com/dyfer/supercollider/blob/topic/scsynth-portaudio-input-output-devices/server/scsynth/SC_PortAudio.cpp#L600) making one last attempt to open default devices. Is that OK?

Reply:
make sense, thanks for the explanation!